File: orchestration/src/tasks/helpers/slam.py
from __future__ import annotations

import json
import os
import shlex
from pathlib import Path
from typing import Any

import tomli_w
from python_on_whales import DockerClient

from src import host
from src.configs.run_config import RunConfig
from src.tasks.helpers.artifacts import file_sha256, write_text
from src.tasks.helpers.navlab_models import (
    remove_container,
)
from src.tasks.helpers.official_stack import (
    build_doctor_summary,
)
from src.tasks.helpers.rosbag_profiles import load_rosbag_metadata_counts

SLAM_BACKEND_CONTAINER = "navlab-p3-slam-backend"
OFFICIAL_IRIS_LIDAR_Z_M = "0.075077"
P3_ROSBAG_CONTAINER = "navlab-p3-rosbag"


def _baseline_env(config: RunConfig) -> dict[str, str]:
    baseline = config.orchestration.official_baseline
    return {
        "DDS_ENABLE": baseline.dds_enable,
        "DDS_DOMAIN_ID": baseline.dds_domain_id,
        "ROS_DOMAIN_ID": baseline.dds_domain_id,
        "RMW_IMPLEMENTATION": baseline.rmw_implementation,
    }


def write_p3_slam_runtime_config(config: RunConfig, path: Path) -> dict[str, Any]:
    p3 = config.orchestration.slam_backend
    data = {
        "slam": {
            "runtime": {
                "backend": p3.backend,
                "use_sim_time": True,
                "launch_package": p3.launch_package,
                "launch_file": p3.launch_file,
                "launch_fake_odom": False,
                "launch_cartographer_backend": True,
                "publish_placeholder_odom": False,
                "cartographer_configuration_basename": p3.cartographer_configuration_basename,
                "imu_source_mode": "topic",
                "imu_source_topic": p3.imu_topic,
                "imu_source_label": "official_gazebo_imu_bridge",
                "imu_min_input_rate_hz": "4.0",
                "require_imu_for_external_nav": False,
                "require_height_for_external_nav": False,
                "external_nav_input_odom_topic": p3.slam_odom_topic,
                "external_nav_output_topic": "/external_nav/odom",
                "external_nav_status_topic": p3.external_nav_status_topic,
                "scan_topic": p3.scan_topic,
                "imu_topic": p3.imu_topic,
                "cartographer_odometry_topic": p3.odometry_topic,
                "odom_topic": p3.slam_odom_topic,
                "slam_status_topic": p3.slam_status_topic,
                "map_frame_id": "map",
                "odom_frame_id": p3.odom_frame_id,
                "base_frame_id": p3.base_frame_id,
                "imu_frame_id": p3.imu_frame_id,
                "laser_frame_id": p3.laser_frame_id,
                "base_frame": p3.base_frame_id,
                "imu_frame": p3.imu_frame_id,
                "laser_frame": p3.laser_frame_id,
                "laser_x": "0",
                "laser_y": "0",
                "laser_z": OFFICIAL_IRIS_LIDAR_Z_M,
                "laser_roll": "0",
                "laser_pitch": "0",
                "laser_yaw": "0",
            }
        }
    }
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(tomli_w.dumps(data).encode("utf-8"))
    return {"path": str(path), "workspace_path": host.workspace_path(path), "sha256": file_sha256(path), "data": data}


def start_p3_slam_container(config: RunConfig, *, runtime_config: Path) -> None:
    remove_container(SLAM_BACKEND_CONTAINER)
    p3 = config.orchestration.slam_backend
    workspace_config = host.workspace_path(runtime_config)
    launch_command = " ".join(
        shlex.quote(arg)
        for arg in [
            "python3",
            "-m",
            "navlab.common.slam.cli",
            "launch",
            "--config",
            workspace_config,
            "--backend",
            p3.backend,
        ]
    )
    DockerClient().run(
        config.slam_image,
        [
            "bash",
            "-lc",
            "source /opt/ros/jazzy/setup.bash && "
            "source /opt/navlab_ws/install/setup.bash && "
            f"exec {launch_command}",
        ],
        detach=True,
        name=SLAM_BACKEND_CONTAINER,
        networks=["host"],
        volumes=[(Path.cwd(), "/workspace")],
        workdir="/workspace",
        user=f"{os.getuid()}:{os.getgid()}",
        envs={
            "SESSION_ID": config.session_id,
            "ROS_DOMAIN_ID": config.orchestration.official_baseline.dds_domain_id,
            "RMW_IMPLEMENTATION": config.orchestration.official_baseline.rmw_implementation,
            "PYTHONPATH": "/workspace",
            "NAVLAB_SLAM_RUNTIME_CONFIG": workspace_config,
        },
    )


def _message_counts(config: RunConfig) -> dict[str, int]:
    metadata = config.artifact_dir / "rosbag" / "metadata.yaml"
    if not metadata.is_file():
        return {}
    return load_rosbag_metadata_counts(metadata)


def collect_odometry_probe(config: RunConfig, *, image: str, topic: str, artifact_name: str) -> dict[str, Any]:
    script = f"""
import json
import math
import time

import rclpy
from nav_msgs.msg import Odometry
from rclpy.qos import qos_profile_sensor_data

topic = {topic!r}
result = {{
    "topic": topic,
    "received": False,
    "message_count": 0,
    "max_jump_m": 0.0,
    "max_yaw_jump_rad": 0.0,
    "stationary_drift_m": 0.0,
}}

def yaw_from_quat(q):
    siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
    cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
    return math.atan2(siny_cosp, cosy_cosp)

def angle_delta(a, b):
    return math.atan2(math.sin(a - b), math.cos(a - b))

rclpy.init()
node = rclpy.create_node("navlab_p3_odometry_probe_" + topic.strip("/").replace("/", "_").replace("-", "_"))
started = time.monotonic()
last_time = None
first_pos = None
prev_pos = None
prev_yaw = None

def callback(msg):
    global first_pos, prev_pos, prev_yaw, last_time
    result["received"] = True
    result["message_count"] += 1
    last_time = time.monotonic()
    pos = msg.pose.pose.position
    q = msg.pose.pose.orientation
    point = (float(pos.x), float(pos.y), float(pos.z))
    yaw = yaw_from_quat(q)
    if first_pos is None:
        first_pos = point
    if prev_pos is not None:
        jump = math.dist(point, prev_pos)
        result["max_jump_m"] = max(float(result["max_jump_m"]), jump)
        result["max_yaw_jump_rad"] = max(float(result["max_yaw_jump_rad"]), abs(angle_delta(yaw, prev_yaw)))
    result["stationary_drift_m"] = max(float(result["stationary_drift_m"]), math.dist(point, first_pos))
    prev_pos = point
    prev_yaw = yaw
    result["frame_id"] = msg.header.frame_id
    result["child_frame_id"] = msg.child_frame_id
    result["latest_position"] = {{"x": point[0], "y": point[1], "z": point[2]}}
    result["latest_yaw_rad"] = yaw

subscription = node.create_subscription(Odometry, topic, callback, qos_profile_sensor_data)
deadline = time.monotonic() + 10.0
while time.monotonic() < deadline:
    rclpy.spin_once(node, timeout_sec=0.1)
result["sample_window_sec"] = time.monotonic() - started
result["rate_hz"] = result["message_count"] / max(result["sample_window_sec"], 0.001)
result["latest_age_sec"] = None if last_time is None else time.monotonic() - last_time
node.destroy_subscription(subscription)
node.destroy_node()
rclpy.shutdown()
print(json.dumps(result, sort_keys=True))
"""
    rc, output = host.docker_run_ros_shell_capture(
        config=config,
        image=image,
        shell_command=f"python3 - <<'PY'\n{script}\nPY",
        name=None,
        network="host",
        envs=_baseline_env(config),
    )
    write_text(config.artifact_dir / artifact_name, output)
    return {"rc": rc, "output": output, "result": _last_json_line(output) if rc == 0 else {}}


def _last_json_line(output: str) -> dict[str, Any]:
    for line in reversed([line.strip() for line in output.splitlines() if line.strip()]):
        try:
            value = json.loads(line)
        except json.JSONDecodeError:
            continue
        return value if isinstance(value, dict) else {}
    return {}


def _float_or_none(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def append_slam_odom_quality_blockers(*, blockers: list[str], p3: Any, slam_odom_result: dict[str, Any]) -> None:
    if not slam_odom_result.get("received"):
        blockers.append(f"P3 did not receive {p3.slam_odom_topic}")
    if slam_odom_result.get("frame_id") != p3.odom_frame_id:
        blockers.append(f"{p3.slam_odom_topic} frame_id is not {p3.odom_frame_id!r}")
    if slam_odom_result.get("child_frame_id") != p3.base_frame_id:
        blockers.append(f"{p3.slam_odom_topic} child_frame_id is not {p3.base_frame_id!r}")
    if float(slam_odom_result.get("rate_hz", 0.0) or 0.0) < p3.min_slam_odom_rate_hz:
        blockers.append("SLAM odom rate is below minimum")
    latest_age = _float_or_none(slam_odom_result.get("latest_age_sec"))
    if latest_age is None or latest_age > p3.max_latest_age_sec:
        blockers.append("SLAM odom latest age is too high")
    max_jump_m = _float_or_none(slam_odom_result.get("max_jump_m"))
    if max_jump_m is None or max_jump_m > p3.max_jump_m:
        blockers.append("SLAM odom jump exceeds threshold")
    max_yaw_jump_rad = _float_or_none(slam_odom_result.get("max_yaw_jump_rad"))
    if max_yaw_jump_rad is None or max_yaw_jump_rad > p3.max_yaw_jump_rad:
        blockers.append("SLAM odom yaw jump exceeds threshold")
    stationary_drift_m = _float_or_none(slam_odom_result.get("stationary_drift_m"))
    if stationary_drift_m is None or stationary_drift_m > p3.max_stationary_drift_m:
        blockers.append("SLAM odom stationary drift exceeds threshold")


def slam_backend_print_command(config: RunConfig, *, runtime_config: Path) -> tuple[int, str]:
    p3 = config.orchestration.slam_backend
    command = " ".join(
        shlex.quote(arg)
        for arg in [
            "python3",
            "-m",
            "navlab.common.slam.cli",
            "print-command",
            "--config",
            host.workspace_path(runtime_config),
            "--backend",
            p3.backend,
        ]
    )
    return host.docker_run_ros_shell_capture(
        config=config,
        image=config.slam_image,
        shell_command=f"source /opt/navlab_ws/install/setup.bash && {command}",
        name=None,
        network=None,
        envs={"NAVLAB_SLAM_RUNTIME_CONFIG": host.workspace_path(runtime_config)},
    )


def _write_foxglove_notes(config: RunConfig) -> None:
    p3 = config.orchestration.slam_backend
    write_text(
        config.artifact_dir / "foxglove_notes.md",
        "\n".join(
            [
                "# NavLab P3 SLAM backend replay notes",
                "",
                "P3 validates the SLAM backend as an observation pipeline. It does not validate hover, path tracking, or ExternalNav control.",
                "",
                "- Fixed frame: `map`.",
                f"- SLAM canonical odom: `{p3.slam_odom_topic}`.",
                f"- SLAM status: `{p3.slam_status_topic}`.",
                f"- Inputs: `{p3.scan_topic}`, `{p3.imu_topic}`, `{p3.odometry_topic}`, `/tf`, `/tf_static`.",
                "- Cartographer map topics: `/map`, `/submap_list`, `/trajectory_node_list`.",
                f"- Diagnostic truth comparison topic: `{p3.truth_diagnostic_topic}`.",
                "- Do not interpret this bag as a flight-complete acceptance. This phase only proves backend quality gates.",
            ]
        )
        + "\n",
    )


def build_p3_doctor_summary(config: RunConfig, *, runtime_config: Path) -> dict[str, Any]:
    p3 = config.orchestration.slam_backend
    blockers: list[str] = []
    baseline_doctor = build_doctor_summary(config)
    if not baseline_doctor.get("ok"):
        blockers.extend(str(item) for item in baseline_doctor.get("blockers", []))
    if p3.uses_gazebo_truth_as_input:
        blockers.append("P3 SLAM backend must not use Gazebo truth as the SLAM odom source")
    if p3.slam_odom_topic in {p3.truth_diagnostic_topic, p3.odometry_topic}:
        blockers.append("P3 canonical SLAM odom topic must be distinct from diagnostic truth / odometry input")
    rc, command = slam_backend_print_command(config, runtime_config=runtime_config)
    if rc != 0:
        blockers.append(f"SLAM backend command could not be built rc={rc}")
    summary = {
        "ok": not blockers,
        "blocked": bool(blockers),
        "blockers": blockers,
        "p3_slam_backend_doctor": {
            "backend": p3.backend,
            "slam_image": config.slam_image,
            "runtime_config": str(runtime_config),
            "runtime_config_sha256": file_sha256(runtime_config) if runtime_config.is_file() else "",
            "command": command.strip(),
            "scan_topic": p3.scan_topic,
            "imu_topic": p3.imu_topic,
            "odometry_topic": p3.odometry_topic,
            "slam_odom_topic": p3.slam_odom_topic,
            "slam_status_topic": p3.slam_status_topic,
            "truth_diagnostic_topic": p3.truth_diagnostic_topic,
            "uses_gazebo_truth_as_input": p3.uses_gazebo_truth_as_input,
            "placeholder_odom_allowed": False,
            "fake_odom_allowed": False,
        },
        "official_baseline_doctor": baseline_doctor,
    }
    return summary

File: orchestration/sim/internal/tasks/helpers/slam.go
package helpers

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	toml "github.com/pelletier/go-toml/v2"
)

const (
	SlamBackendContainer                 = "navlab-slam-backend"
	OfficialExternalNavBridgeParams      = "/opt/navlab_ws/install/navlab_external_nav_bridge/share/navlab_external_nav_bridge/config/navlab_external_nav_bridge.params.yaml"
	OfficialIrisLidarZM                  = "0.075077"
	CartographerOdometryInputTopic       = "/cartographer/odometry_input"
	CartographerTFTopic                  = "/navlab/slam/tf"
	DiagnosticTruthOdometryTopic         = "/odometry"
	DiagnosticGazeboModelOdometryTopic   = "/gazebo/model/odometry"
	DiagnosticGazeboTFTopic              = "/gazebo/tf"
	DiagnosticGazeboTFStaticTopic        = "/gazebo/tf_static"
	DiagnosticCartographerConfigBasename = "navlab_cartographer_2d_diagnostic_odom.lua"
	HoverCartographerConfigBasename      = "navlab_cartographer_2d_hover.lua"
	OfficialCartographerConfigDir        = "/opt/navlab_ws/install/navlab_cartographer_adapter/share/navlab_cartographer_adapter/config"
)

type SlamRuntimeSpec struct {
	Backend                            string
	UseSimTime                         bool
	LaunchPackage                      string
	LaunchFile                         string
	LaunchCartographerBackend          bool
	CartographerConfigurationDirectory string
	CartographerConfigurationBasename  string
	ScanTopic                          string
	IMUSourceTopic                     string
	IMUTopic                           string
	OdometryTopic                      string
	CartographerTFTopic                string
	PublishGlobalTF                    bool
	GlobalTFTopic                      string
	OdomSourceMode                     string
	SlamOdomTopic                      string
	ExternalNavInputOdomTopic          string
	SlamStatusTopic                    string
	ExternalNavStatusTopic             string
	MapFrameID                         string
	OdomFrameID                        string
	BaseFrameID                        string
	IMUFrameID                         string
	LaserFrameID                       string
	RequireIMUForQuality               bool
	RequireScanForQuality              bool
	LowObservabilityMode               bool
}

func ExternalNavBridgeParamsOverride(spec SlamRuntimeSpec) string {
	if spec.Backend == "" {
		spec = DefaultSlamRuntimeSpec()
	}
	externalNavInputOdomTopic := spec.ExternalNavInputOdomTopic
	if externalNavInputOdomTopic == "" {
		externalNavInputOdomTopic = spec.SlamOdomTopic
	}
	return fmt.Sprintf(`navlab_external_nav_bridge_node:
  ros__parameters:
    odom_timeout_ms: 500
    imu_timeout_ms: 500
    scan_timeout_ms: 1000
    height_timeout_ms: 2000
    require_imu_for_output: false
    require_height_for_output: true
    require_imu_for_quality: %t
    require_scan_for_quality: %t
    slam_quality_gate_enabled: true
    low_observability_mode: %t
    input_odom_topic: %s
    imu_topic: %s
    scan_topic: %s
    height_topic: /height/estimate
    max_height_covariance: 4.0
    min_odom_rate_hz: 4.0
    min_imu_rate_hz: 4.0
    min_scan_rate_hz: 2.0
    max_position_jump_m: 0.75
    max_yaw_jump_rad: 0.75
    jump_hold_ms: 2000
    min_observable_horizontal_span_m: 0.10
    min_scan_valid_ratio_for_quality: 0.50
    min_scan_hit_ratio_for_quality: 0.25
    min_scan_range_span_m_for_quality: 1.0
    min_scan_range_stddev_m_for_quality: 0.20
    min_scan_observed_quadrants_for_quality: 3
    scan_max_range_hit_margin_m: 0.05
    output_odom_topic: /external_nav/odom
    status_topic: %s
    output_frame_id: external_nav
    output_child_frame_id: %s
    ap_tf_topic: /ap/tf
    ap_tf_parent_frame: %s
    ap_tf_child_frame: %s
    expected_odom_frame_id: %s
    expected_odom_child_frame_id: %s
    coordinate_mode: pass_through_enu_flu
`,
		spec.RequireIMUForQuality,
		spec.RequireScanForQuality,
		spec.LowObservabilityMode,
		externalNavInputOdomTopic,
		spec.IMUTopic,
		spec.ScanTopic,
		spec.ExternalNavStatusTopic,
		spec.BaseFrameID,
		spec.OdomFrameID,
		spec.BaseFrameID,
		spec.MapFrameID,
		spec.BaseFrameID,
	)
}

type SlamContainerSpec struct {
	Image                  string
	RuntimeConfigPath      string
	Backend                string
	SessionID              string
	RosDomainID            string
	RMWImplementation      string
	WorkspaceHostPath      string
	WorkspaceContainerPath string
	User                   string
}

func DefaultSlamRuntimeSpec() SlamRuntimeSpec {
	return SlamRuntimeSpec{
		Backend:                            "cartographer",
		UseSimTime:                         true,
		LaunchPackage:                      "navlab_slam_bringup",
		LaunchFile:                         "navlab_slam_bringup.launch.py",
		LaunchCartographerBackend:          true,
		CartographerConfigurationDirectory: "",
		CartographerConfigurationBasename:  "navlab_cartographer_2d_real.lua",
		ScanTopic:                          "/scan",
		IMUSourceTopic:                     "/imu",
		IMUTopic:                           "/imu",
		OdometryTopic:                      CartographerOdometryInputTopic,
		CartographerTFTopic:                CartographerTFTopic,
		PublishGlobalTF:                    true,
		GlobalTFTopic:                      "/tf",
		OdomSourceMode:                     "slam_tf",
		SlamOdomTopic:                      "/slam/odom",
		ExternalNavInputOdomTopic:          "/slam/odom",
		SlamStatusTopic:                    "/navlab/slam/status",
		ExternalNavStatusTopic:             "/external_nav/status",
		MapFrameID:                         "map",
		OdomFrameID:                        "odom",
		BaseFrameID:                        "base_link",
		IMUFrameID:                         "imu_link",
		LaserFrameID:                       "base_scan",
	}
}

func HoverSlamRuntimeSpec(spec SlamRuntimeSpec) SlamRuntimeSpec {
	if spec.Backend == "" {
		spec = DefaultSlamRuntimeSpec()
	}
	spec.CartographerConfigurationBasename = HoverCartographerConfigBasename
	spec.IMUSourceTopic = "/imu"
	spec.IMUTopic = "/navlab/slam/imu"
	spec.OdometryTopic = CartographerOdometryInputTopic
	return spec
}

func WriteSlamRuntimeConfig(path string, spec SlamRuntimeSpec) error {
	if spec.Backend == "" {
		spec = DefaultSlamRuntimeSpec()
	}
	externalNavInputOdomTopic := spec.ExternalNavInputOdomTopic
	if externalNavInputOdomTopic == "" {
		externalNavInputOdomTopic = spec.SlamOdomTopic
	}
	imuSourceTopic := spec.IMUSourceTopic
	if imuSourceTopic == "" {
		imuSourceTopic = spec.IMUTopic
	}
	data := map[string]any{
		"slam": map[string]any{
			"runtime": map[string]any{
				"backend":                                   spec.Backend,
				"use_sim_time":                              spec.UseSimTime,
				"launch_package":                            spec.LaunchPackage,
				"launch_file":                               spec.LaunchFile,
				"launch_fake_odom":                          false,
				"launch_cartographer_backend":               spec.LaunchCartographerBackend,
				"publish_placeholder_odom":                  false,
				"cartographer_configuration_directory":      spec.CartographerConfigurationDirectory,
				"cartographer_configuration_basename":       spec.CartographerConfigurationBasename,
				"imu_source_mode":                           "topic",
				"imu_source_topic":                          imuSourceTopic,
				"imu_source_label":                          "official_gazebo_imu_bridge",
				"imu_min_input_rate_hz":                     "4.0",
				"require_imu_for_external_nav":              false,
				"require_height_for_external_nav":           true,
				"external_nav_input_odom_topic":             externalNavInputOdomTopic,
				"external_nav_expected_odom_frame_id":       spec.MapFrameID,
				"external_nav_expected_odom_child_frame_id": spec.BaseFrameID,
				"external_nav_output_topic":                 "/external_nav/odom",
				"external_nav_status_topic":                 spec.ExternalNavStatusTopic,
				"scan_topic":                                spec.ScanTopic,
				"imu_topic":                                 spec.IMUTopic,
				"cartographer_odometry_topic":               spec.OdometryTopic,
				"cartographer_tf_topic":                     spec.CartographerTFTopic,
				"publish_global_tf":                         spec.PublishGlobalTF,
				"global_tf_topic":                           spec.GlobalTFTopic,
				"cached_odom_publish_rate_hz":               "10.0",
				"odom_source_mode":                          spec.OdomSourceMode,
				"odom_topic":                                spec.SlamOdomTopic,
				"slam_status_topic":                         spec.SlamStatusTopic,
				"map_frame_id":                              spec.MapFrameID,
				"odom_frame_id":                             spec.OdomFrameID,
				"base_frame_id":                             spec.BaseFrameID,
				"imu_frame_id":                              spec.IMUFrameID,
				"laser_frame_id":                            spec.LaserFrameID,
				"base_frame":                                spec.BaseFrameID,
				"imu_frame":                                 spec.IMUFrameID,
				"laser_frame":                               spec.LaserFrameID,
				"laser_x":                                   "0",
				"laser_y":                                   "0",
				"laser_z":                                   OfficialIrisLidarZM,
				"laser_roll":                                "0",
				"laser_pitch":                               "0",
				"laser_yaw":                                 "0",
			},
		},
	}
	encoded, err := toml.Marshal(data)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, encoded, 0o644)
}

func SlamDockerArgs(spec SlamContainerSpec) ([]string, error) {
	if spec.Image == "" {
		return nil, fmt.Errorf("slam image is required")
	}
	workspaceHostPath := spec.WorkspaceHostPath
	if workspaceHostPath == "" {
		workspaceHostPath = "."
	}
	workspaceContainerPath := spec.WorkspaceContainerPath
	if workspaceContainerPath == "" {
		workspaceContainerPath = "/workspace"
	}
	backend := spec.Backend
	if backend == "" {
		backend = "cartographer"
	}
	command := "source /opt/ros/${ROS_DISTRO:-humble}/setup.bash && " +
		"source /opt/navlab_ws/install/setup.bash && " +
		"exec python3 -m navlab.common.slam.cli launch --config " + shellQuote(spec.RuntimeConfigPath) +
		" --backend " + shellQuote(backend)
	args := []string{
		"run",
		"--detach",
		"--name", SlamBackendContainer,
		"--network", "host",
		"--volume", workspaceHostPath + ":" + workspaceContainerPath,
		"--workdir", workspaceContainerPath,
		"--env", "SESSION_ID=" + spec.SessionID,
		"--env", "ROS_DOMAIN_ID=" + spec.RosDomainID,
		"--env", "RMW_IMPLEMENTATION=" + spec.RMWImplementation,
		"--env", "PYTHONPATH=" + workspaceContainerPath,
		"--env", "NAVLAB_SLAM_RUNTIME_CONFIG=" + spec.RuntimeConfigPath,
	}
	if spec.User != "" {
		args = append(args, "--user", spec.User)
	}
	args = append(args, spec.Image, "bash", "-lc", command)
	return args, nil
}

func StartSlamContainer(ctx context.Context, runner CommandRunner, spec SlamContainerSpec) error {
	if runner == nil {
		runner = ExecCommandRunner{}
	}
	_ = RemoveContainer(ctx, runner, SlamBackendContainer)
	args, err := SlamDockerArgs(spec)
	if err != nil {
		return err
	}
	_, err = runner.Run(ctx, "docker", args...)
	return err
}
